src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs

```diff
@@ -3,86 +3,89 @@
-                    // We're performing writes, so we need to enqueue this work to the exclusive
-                    // scheduler.  Because this happens in the future, we may end up getting run
-                    // after the storage system has shutdown.  So we need to check to make sure we
-                    // should still continue once we wake up.
-                    var token = _shutdownTokenSource.Token;
-                    _flushTask = Task.Delay(FlushAllDelayMS, token).ContinueWith(
-                        s_flushInMemoryDataToDiskIfNotShutdown, this, token, TaskContinuationOptions.None, _readerWriterLock.ExclusiveScheduler);
-                }
-            }
+            // actual value isn't relevant.  this just ensures that a flush will happen in the future.
+            _flushQueue.AddWork(true);
         }
 
-        private static readonly Action<Task, object> s_flushInMemoryDataToDiskIfNotShutdown = (Task _1, object self) =>
+        private Task FlushInMemoryDataToDiskIfNotShutdownAsync(ImmutableArray<bool> _, CancellationToken cancellationToken)
         {
-            var storage = (SQLitePersistentStorage)self;
+            // When we are asked to flush, go actually acquire the write-scheduler and perform the actual writes from
+            // it. Note: this is only called max every FlushAllDelayMS.  So we don't bother trying to avoid the delegate
+            // allocation here.
+            return this.PerformWriteAsync(FlushInMemoryDataToDisk, cancellationToken);
+        }
 
-            lock (storage._flushGate)
-            {
-                // Indicate that there is no outstanding write task.  The next request to 
-                // write will cause one to be kicked off.
-                storage._flushTask = null;
+        private void FlushWritesOnClose()
+        {
+            // Issue a write task to write this all out to disk.
+            //
+            // Note: this only happens on close, so we don't try to avoid allocations here.
 
-                if (storage._shutdownTokenSource.IsCancellationRequested)
+            var writeTask = PerformWriteAsync(
+                () =>
                 {
-                    // Don't flush from a bg task if we've been asked to shutdown.  The shutdown
-                    // logic in the storage service will take care of the final writes to the main
-                    // db.
-                    return;
-                }
+                    // Perform the actual write while having exclusive access to the scheduler.
+                    FlushInMemoryDataToDisk();
 
-                // Haven't been shutdown.  Actually go and move any outstanding data to the real DB.
-                using var _ = storage.GetPooledConnection(out var connection);
-                storage.FlushInMemoryDataToDisk_MustRunUnderLock(connection);
-            }
-        };
+                    // Now that we've done this, definitely cancel any further work. From this point on, it is now
```

Comment:
The idea now I guess is since this is being cancelled while under the write lock nothing else can sneak in between the call to FlushInMemoryDataToDisk() and the Cancel()?

Reply:
that's exactly it.  it's very nice and clean IMO.  all flushes are serialized.  so they always see a consistent state of cancellation.  as mentioned in the comment, either:

1. Dispose/Close happens first. in which case it flushes the remaining work, cancels the token and then returns.  The next bg flush task wake up, sees hte token, and bails.  No work is missed because it's not legal for a client to somehow read/write after dispose.
2. the bg flush task happens first.  is sees it isn't canceled, writes out and retursn.  the dispose/close task tehn runs and flushes any final work that was added pre-dispose/close.